panic: first replica is not the primary replica for the token

Please answer these questions before submitting your issue. Thanks!

### What version of Cassandra are you using?

[cqlsh 5.0.1 | Cassandra 3.0.14.32 | CQL spec 3.4.0 | Native protocol v4]

keyspace replica: replication = {'class': 'NetworkTopologyStrategy', 'dca1': '2'}

### What version of Gocql are you using?

revision = "16cf9ea1b3e28090d416d36528f304e4553e6b56"

### What did you do?

Change DCAwareRoundRobinPolicy to TokenAwareHostPolicy(DCAwareRoundRobinPolicy)

### What did you expect to see?
Correctly initialized session for read/write

### What did you see instead?

This is different from the similar issue I found from this repo, there is no node add/remove, when we deploy the change and try to create session, the exception was thrown.

panic: first replica is not the primary replica for the token

goroutine 971 [running]:
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.(*networkTopology).replicaMap(0xc01c5445b8, 0xc010df5b00, 0x83, 0x83, 0xc030916000, 0x8300, 0x9d55, 0xc000bf2000)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/topology.go:203 +0x1366
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).updateReplicas(0xc00257e000, 0xc002862310, 0xc0015a2320, 0xe)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/policies.go:468 +0x2b9
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.(*tokenAwareHostPolicy).KeyspaceChanged(0xc00257e000, 0xc0015a2320, 0xe, 0x0, 0x0)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/policies.go:452 +0x9a
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.(*Session).init(0xc008536e00, 0xc00257e070, 0x0)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/session.go:256 +0x3cb
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.NewSession(0xc00274f000, 0x41, 0x80, 0x167581a, 0x5, 0x4, 0x2540be400, 0x0, 0x6b4c, 0xc0015a2320, ...)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/session.go:161 +0x824
code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql.(*ClusterConfig).CreateSession(...)
	code.uber.internal/infra/jaeger/vendor/github.com/gocql/gocql/cluster.go:185
code.uber.internal/infra/jaeger/cmd/ingester/app/cassandra.NewSession.func1(0xc0000d3f80, 0x16b4879, 0x2e, 0x0)
---

If you are having connectivity related issues please share the following additional information

### Describe your Cassandra cluster
please provide the following information

- output of `nodetool status`
- output of `SELECT peer, rpc_address FROM system.peers`
- rebuild your application with the `gocql_debug` tag and post the output

I have similar issue but with two DC and keyspace with replication only on single DC:
For example:
```go
package main

import (
	"log"

	"github.com/gocql/gocql"
)

func main() {
	cluster := gocql.NewCluster("127.0.1.1")
	cluster.Keyspace = "testkv"
	rr := gocql.DCAwareRoundRobinPolicy("dc1")
	cluster.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(rr)
	session, err := cluster.CreateSession()
	if err != nil {
		log.Fatal(err)
	}
	defer session.Close()
}
```

With two DC running on localhost (of cause I have an issue in production. But it is easily reproduced in such simple local setup).
```
$ nodetool status
Datacenter: dc1
===============
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address    Load       Tokens       Owns (effective)  Host ID                               Rack
UN  127.0.1.1  179,66 KiB  256          100,0%            5ff09954-d433-462e-9c00-98a07868142d  rack1
Datacenter: dc2
===============
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address    Load       Tokens       Owns (effective)  Host ID                               Rack
UN  127.0.1.2  189,7 KiB  256          0,0%              49614024-f089-40b0-88ab-104ebab706d4  rack1
```

And keyspace:
```
cqlsh> describe testkv;             
                                                         
CREATE KEYSPACE testkv WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '1'}  AND durable_writes = true;
```

And panic stacktrace is same to above:
```
panic: first replica is not the primary replica for the token: expected 127.0.1.1 got 127.0.1.2

goroutine 1 [running]:
github.com/gocql/gocql.(*networkTopology).replicaMap(0xc000196078, 0xc000096380, 0xc000196078, 0x0, 0x0)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/topology.go:261 +0x17d1
github.com/gocql/gocql.(*tokenAwareHostPolicy).updateReplicas(0xc0001383f0, 0xc000012b80, 0x706d53, 0x6)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/policies.go:449 +0x25f
github.com/gocql/gocql.(*tokenAwareHostPolicy).KeyspaceChanged(0xc0001383f0, 0x706d53, 0x6, 0x0, 0x0)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/policies.go:434 +0x99
github.com/gocql/gocql.(*Session).init(0xc000122380, 0xc000142180, 0x0)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/session.go:280 +0x5ca
github.com/gocql/gocql.NewSession(0xc000012960, 0x1, 0x1, 0x706956, 0x5, 0x0, 0x23c34600, 0x23c34600, 0x2352, 0x706d53, ...)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/session.go:166 +0x809
github.com/gocql/gocql.(*ClusterConfig).CreateSession(...)
        /home/yura/go/pkg/mod/github.com/gocql/gocql@v0.0.0-20200228163523-cd4b606dd2fb/cluster.go:194
main.main()
        /home/archive/cassandra_cluster/test_app/main.go:14 +0x2aa
```
As you can see in stack trace folders, gocql is the latest commit (but from 20190927095247-bd5f930c6137 fails in similar way).

When keyspace replication is set to both DC (i.e. with `{'class': 'NetworkTopologyStrategy', 'dc1': '1', 'dc2': '1'}`), it doesn't fail. Unfortunatelly there is no way to add DC in production without being in situation when replication is set to single DC.

--- a/AUTHORS
+++ b/AUTHORS
@@ -113,3 +113,4 @@ Thomas Meson <zllak@hycik.org>
 Martin Sucha <martin.sucha@kiwi.com>; <git@mm.ms47.eu>
 Pavel Buchinchik <p.buchinchik@gmail.com>
 Rintaro Okamura <rintaro.okamura@gmail.com>
+Yura Sokolov <y.sokolov@joom.com>; <funny.falcon@gmail.com>

--- a/topology.go
+++ b/topology.go
@@ -171,7 +171,7 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 	}
 
 	tokens := tokenRing.tokens
-	replicaRing := make(tokenRingReplicas, len(tokens))
+	replicaRing := make(tokenRingReplicas, 0, len(tokens))
 
 	var totalRF int
 	for _, rf := range n.dcs {
@@ -179,6 +179,11 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 	}
 
 	for i, th := range tokenRing.tokens {
+		if rf := n.dcs[th.host.DataCenter()]; rf == 0 {
+			// skip this token since no replica in this datacenter.
+			continue
+		}
+
 		for k, v := range skipped {
 			skipped[k] = v[:0]
 		}
@@ -202,9 +207,9 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 			dc := h.DataCenter()
 			rack := h.Rack()
 
-			rf, ok := n.dcs[dc]
-			if !ok {
-				// skip this DC, dont know about it
+			rf := n.dcs[dc]
+			if rf == 0 {
+				// skip this DC, dont know about it or replication factor is zero
 				continue
 			} else if replicasInDC[dc] >= rf {
 				if replicasInDC[dc] > rf {
@@ -261,10 +266,10 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 			panic(fmt.Sprintf("first replica is not the primary replica for the token: expected %v got %v", replicas[0].ConnectAddress(), th.host.ConnectAddress()))
 		}
 
-		replicaRing[i] = hostTokens{th.token, replicas}
+		replicaRing = append(replicaRing, hostTokens{th.token, replicas})
 	}
 
-	if len(replicaRing) != len(tokens) {
+	if len(n.dcs) == len(dcRacks) && len(replicaRing) != len(tokens) {
 		panic(fmt.Sprintf("token map different size to token ring: got %d expected %d", len(replicaRing), len(tokens)))
 	}
 

--- a/topology_test.go
+++ b/topology_test.go
@@ -90,76 +90,86 @@ func TestPlacementStrategy_NetworkStrategy(t *testing.T) {
 	}
 	sort.Sort(&tokenRing{tokens: tokens})
 
-	strat := &networkTopology{
-		dcs: map[string]int{
-			"dc1": 1,
-			"dc2": 2,
-			"dc3": 3,
+	strats := []*networkTopology{
+		&networkTopology{
+			dcs: map[string]int{
+				"dc1": 1,
+				"dc2": 2,
+				"dc3": 3,
+			},
+		},
+		&networkTopology{
+			dcs: map[string]int{
+				"dc2": 2,
+				"dc3": 3,
+			},
 		},
 	}
 
-	var expReplicas int
-	for _, rf := range strat.dcs {
-		expReplicas += rf
-	}
-
-	tokenReplicas := strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens})
-	if len(tokenReplicas) != len(tokens) {
-		t.Fatalf("expected replica map to have %d items but has %d", len(tokens), len(tokenReplicas))
-	}
-	if !sort.IsSorted(tokenReplicas) {
-		t.Fatal("replica map was not sorted by token")
-	}
-
-	for token, replicas := range tokenReplicas {
-		if len(replicas.hosts) != expReplicas {
-			t.Fatalf("expected to have %d replicas got %d for token=%v", expReplicas, len(replicas.hosts), token)
+	for _, strat := range strats {
+		var expReplicas int
+		for _, rf := range strat.dcs {
+			expReplicas += rf
 		}
-	}
 
-	for dc, rf := range strat.dcs {
-		dcTokens := dcRing[dc]
-		for i, th := range dcTokens {
-			token := th.token
-			allReplicas := tokenReplicas.replicasFor(token)
-			if allReplicas.token != token {
-				t.Fatalf("token %v not in replica map", token)
+		tokenReplicas := strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens})
+		if needTokens := hostsPerDC * len(strat.dcs); len(tokenReplicas) != needTokens {
+			t.Fatalf("expected replica map to have %d items but has %d", needTokens, len(tokenReplicas))
+		}
+		if !sort.IsSorted(tokenReplicas) {
+			t.Fatal("replica map was not sorted by token")
+		}
+
+		for token, replicas := range tokenReplicas {
+			if len(replicas.hosts) != expReplicas {
+				t.Fatalf("expected to have %d replicas got %d for token=%v", expReplicas, len(replicas.hosts), token)
 			}
+		}
 
-			var replicas []*HostInfo
-			for _, replica := range allReplicas.hosts {
-				if replica.dataCenter == dc {
-					replicas = append(replicas, replica)
+		for dc, rf := range strat.dcs {
+			dcTokens := dcRing[dc]
+			for i, th := range dcTokens {
+				token := th.token
+				allReplicas := tokenReplicas.replicasFor(token)
+				if allReplicas.token != token {
+					t.Fatalf("token %v not in replica map", token)
 				}
-			}
 
-			if len(replicas) != rf {
-				t.Fatalf("expected %d replicas in dc %q got %d", rf, dc, len(replicas))
-			}
+				var replicas []*HostInfo
+				for _, replica := range allReplicas.hosts {
+					if replica.dataCenter == dc {
+						replicas = append(replicas, replica)
+					}
+				}
 
-			var lastRack string
-			for j, replica := range replicas {
-				// expected is in the next rack
-				var exp *HostInfo
-				if lastRack == "" {
-					// primary, first replica
-					exp = dcTokens[(i+j)%len(dcTokens)].host
-				} else {
-					for k := 0; k < len(dcTokens); k++ {
-						// walk around the ring from i + j to find the next host the
-						// next rack
-						p := (i + j + k) % len(dcTokens)
-						h := dcTokens[p].host
-						if h.rack != lastRack {
-							exp = h
-							break
+				if len(replicas) != rf {
+					t.Fatalf("expected %d replicas in dc %q got %d", rf, dc, len(replicas))
+				}
+
+				var lastRack string
+				for j, replica := range replicas {
+					// expected is in the next rack
+					var exp *HostInfo
+					if lastRack == "" {
+						// primary, first replica
+						exp = dcTokens[(i+j)%len(dcTokens)].host
+					} else {
+						for k := 0; k < len(dcTokens); k++ {
+							// walk around the ring from i + j to find the next host the
+							// next rack
+							p := (i + j + k) % len(dcTokens)
+							h := dcTokens[p].host
+							if h.rack != lastRack {
+								exp = h
+								break
+							}
+						}
+						if exp.rack == lastRack {
+							panic("no more racks")
 						}
 					}
-					if exp.rack == lastRack {
-						panic("no more racks")
-					}
+					lastRack = replica.rack
 				}
-				lastRack = replica.rack
 			}
 		}
 	}

